src/derivative_operators/concretization.jl

```diff
@@ -163,57 +163,194 @@ function Base.convert(::Type{AbstractMatrix},A::AbstractBC{T}) where T
+
+    interior = CartesianIndices(Tuple(ranges))
+    I1 = CartesianIndex(Tuple(ones(Int64, N)))
+    for I in interior
+        i = c2l(I, s_pad)
+        j = c2l(I-ē[D], s)
+        QL[i,j] = one(T)
+    end
+    ranges[D] = 1
+    lower = CartesianIndices((Tuple(ranges)))
+    ranges[D] = s_pad[D]
+    upper = CartesianIndices((Tuple(ranges)))
+    for K in CartesianIndices(upper)
+        I = CartesianIndex(Tuple(K)[setdiff(1:N, D)])
+        il = c2l(lower[K], s_pad)
+        iu = c2l(upper[K], s_pad)
+        Qb[il] = Q[2][I][1]
+        Qb[iu] = Q[2][I][2]
+        for k in 0:s[D]-1
+            j = c2l(lower[K] + k*ē[D], s)
+            QL[il, j] = Q[1][I][1][k+1]
+            QL[iu, j] = Q[1][I][2][k+1]
+        end
+    end
+
+    return (QL, Qb)
+end
 
 """
-Returns a tuple, the first element of which is an array of the shape of the boundary,
-filled with the linear operator parts of the respective Atomic BCs.
-the second element is a similarly sized array of the affine parts.
+This is confusing, but it does work
 """
-function LinearAlgebra.Array(Q::MultiDimDirectionalBC{T, B, D, N, K}, M) where {T, B, D,N,K}
-    bc_tuples = Array.(Q.BCs, fill(M, size(Q.BCs)))
-    Q_L = [bc_tuple[1] for bc_tuple in bc_tuples]
-    inds = Array(1:N)
-    inds[1], inds[D] = inds[D], inds[1]
-    Q_b = [permutedims(add_dims(bc_tuple[2], N-1), inds) for bc_tuple in bc_tuples]
+function LinearAlgebra.Array(Q::ComposedMultiDimBC{T, B, N,M} , s::NTuple{N,G}) where {T, B, N, M, G<:Int}
+    for d in 1:N
+        @assert size(Q.BCs[d]) == perpindex(s, d) "The size of the BC array in Q along dimension $d, $(size(Q.BCs[d])) is incompatible with s, $s"
+    end
+    s_pad = s.+2
+    Q = Tuple(_concretize.(Q.BCs, s)) #essentially finding the first and last rows of the matrix part and affine part for every atomic BC
 
-    return (Q_L, Q_b)
+    QL = zeros(T, prod(s_pad), prod(s))
+    Qb = zeros(T, prod(s_pad))
+
+    ranges = Union{typeof(1:10), Int64}[2:s_pad[i]-1 for i in 1:N] #Set up indices corresponding to the interior
+    interior = CartesianIndices(Tuple(ranges))
+
+    ē = unit_indices(N) #setup unit indices in each direction
+    I1 = CartesianIndex(Tuple(ones(Int64, N))) #setup the ones index
+    for I in interior #loop over interior
+        i = c2l(I, s_pad) #find the index on the padded side
+        j = c2l(I-I1, s)  #find the index on the unpadded side
+        QL[i,j] = one(T)  #create a padded identity matrix
+    end
+    for dim in 1:N #Loop over boundaries
+        r_ = deepcopy(ranges)
+        r_[dim] = 1
+        lower = CartesianIndices((Tuple(r_))) #set up upper anmd lower indices
+        r_[dim] = s_pad[dim]
+        upper = CartesianIndices((Tuple(r_)))
+        for K in CartesianIndices(upper) #for every element of the boundaries
+            I = CartesianIndex(Tuple(K)[setdiff(1:N, dim)]) #convert K to 2D index for indexing the BC arrays
+            il = c2l(lower[K], s_pad) #Translate to linear indices

```

Comment:
c21 looks like it could have a better name, and why would a user want to use this? It seems internal.